// firebase-function/lib/src/crypter/ICrypter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ICrypter = void 0;
const bytesCoder_1 = require("../bytesCoder");
class ICrypter {
    padding;
    constructor(padding) {
        this.padding = padding;
    }
    encrypt(data) {
        const paddedData = this.padding.addPadding(data, this.blockSize);
        return this.encryptBlocks(paddedData);
    }
    decrypt(data) {
        const decryptedData = this.decryptBlocks(data);
        return this.padding.removePadding(decryptedData);
    }
    encodeAndEncrypt(value) {
        const bytesCoder = new bytesCoder_1.Utf8BytesCoder();
        const jsonString = value === undefined ? '' : JSON.stringify(value);
        const data = bytesCoder.encode(jsonString);
        return this.encrypt(data);
    }
    decryptAndDecode(data) {
        const decryptedData = this.decrypt(data);
        const bytesCoder = new bytesCoder_1.Utf8BytesCoder();
        const value = bytesCoder.decode(decryptedData);
        return (value === '' ? undefined : JSON.parse(value));
    }
}
exports.ICrypter = ICrypter;
