// firebase-function/lib/test/parameter/ParameterContainer.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const src_1 = require("../../src");
const testSrc_1 = require("../../src/testSrc");
describe('ParameterContainer', () => {
    const logger = new src_1.VoidLogger();
    it('get invalid parameter', () => {
        const parameterContainer = new src_1.ParameterContainer({
            // eslint-disable-next-line
            value1: undefined,
            value2: null,
            value3: 'asdf'
        }, logger.nextIndent);
        (0, testSrc_1.expect)(() => parameterContainer.parameter('value0', new src_1.ValueParameterBuilder('number'))).to.throw();
        (0, testSrc_1.expect)(() => parameterContainer.parameter('value1', new src_1.ValueParameterBuilder('number'))).to.throw();
        (0, testSrc_1.expect)(() => parameterContainer.parameter('value2', new src_1.ValueParameterBuilder('number'))).to.throw();
        (0, testSrc_1.expect)(() => parameterContainer.parameter('value3', new src_1.ValueParameterBuilder('number'))).to.throw();
    });
    it('get parameter', () => {
        const parameterContainer = new src_1.ParameterContainer({
            value0: undefined,
            value1: true,
            value2: 'asdf',
            value3: 12,
            value5: { value: 'asdf' },
            value6: null
        }, logger.nextIndent);
        // eslint-disable-next-line @typescript-eslint/no-confusing-void-expression
        (0, testSrc_1.expect)(parameterContainer.parameter('value0', new src_1.ValueParameterBuilder('undefined'))).to.be.equal(undefined);
        (0, testSrc_1.expect)(parameterContainer.parameter('value1', new src_1.ValueParameterBuilder('boolean'))).to.be.equal(true);
        (0, testSrc_1.expect)(parameterContainer.parameter('value2', new src_1.ValueParameterBuilder('string'))).to.be.equal('asdf');
        (0, testSrc_1.expect)(parameterContainer.parameter('value3', new src_1.ValueParameterBuilder('number'))).to.be.equal(12);
        (0, testSrc_1.expect)(parameterContainer.parameter('value5', new src_1.ValueParameterBuilder('object'))).to.be.deep.equal({ value: 'asdf' });
        (0, testSrc_1.expect)(parameterContainer.parameter('value6', new src_1.ValueParameterBuilder('object'))).to.be.equal(null);
    });
});
