// firebase-function/lib/src/logger/LoggingProperty.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LoggingProperty = void 0;
const StringBuilder_1 = require("../utils/StringBuilder");
const util_1 = require("util");
class LoggingProperty {
    functionName;
    logLevel;
    indent;
    details;
    constructor(functionName, logLevel, indent, details) {
        this.functionName = functionName;
        this.logLevel = logLevel;
        this.indent = indent;
        this.details = details;
    }
    completeLog(verbose) {
        const builder = new StringBuilder_1.StringBuilder();
        builder.appendLine(`${this.prefix(true)}[${this.logLevel}: ${this.functionName}]${verbose ? ': {' : ''}${!verbose || this.details ? '' : '}'}`);
        if (this.details && verbose) {
            for (const entry of Object.entries(this.details))
                builder.append(this.detailString(entry[0], entry[1]));
            builder.appendLine(`${this.prefix()}}`);
        }
        return builder.toString();
    }
    detailString(key, detail) {
        const builder = new StringBuilder_1.StringBuilder();
        const jsonString = (0, util_1.inspect)(detail, { compact: true, depth: null, maxArrayLength: 25, maxStringLength: 250, breakLength: Number.POSITIVE_INFINITY });
        builder.appendLine(`${this.prefix()}    ${key}: ${jsonString}`);
        return builder.toString();
    }
    prefix(start = false) {
        return '|   '.repeat(this.indent) + (start ? '> ' : '| ');
    }
}
exports.LoggingProperty = LoggingProperty;
