// firebase-function/lib/src/logger/Logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Logger = void 0;
const LoggingProperty_1 = require("./LoggingProperty");
const StringBuilder_1 = require("../utils/StringBuilder");
class Logger {
    verbose;
    properties;
    currentIndent;
    constructor(verbose, properties, currentIndent = 0) {
        this.verbose = verbose;
        this.properties = properties;
        this.currentIndent = currentIndent;
    }
    get nextIndent() {
        return new Logger(this.verbose, this.properties, this.currentIndent + 1);
    }
    get completeLog() {
        const builder = new StringBuilder_1.StringBuilder();
        for (const property of this.properties)
            builder.append(property.completeLog(this.verbose));
        return builder.toString();
    }
    static start(functionName, details = null, logLevel = 'debug', verbose = false) {
        const property = new LoggingProperty_1.LoggingProperty(functionName, logLevel, 0, details);
        return new Logger(verbose, [property]);
    }
    log(functionName, details = null, logLevel = 'debug') {
        this.properties.push(new LoggingProperty_1.LoggingProperty(functionName, logLevel, this.currentIndent, details));
    }
}
exports.Logger = Logger;
