// firebase-function/lib/src/utils/Dictionary.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DictionaryTypeBuilder = exports.Dictionary = void 0;
const Flattable_1 = require("./Flattable");
const record_1 = require("./record");
class Dictionary {
    dictionary;
    constructor(dictionary = {}) {
        this.dictionary = dictionary;
    }
    get(key) {
        return this.dictionary[Flattable_1.Flattable.flatten(key)];
    }
    set(key, value) {
        this.dictionary[Flattable_1.Flattable.flatten(key)] = value;
    }
    has(key) {
        return Flattable_1.Flattable.flatten(key) in this.dictionary;
    }
    get keys() {
        return (0, record_1.keys)(this.dictionary);
    }
    get values() {
        return (0, record_1.values)(this.dictionary);
    }
    get entries() {
        return (0, record_1.entries)(this.dictionary);
    }
    map(callbackFn) {
        return new Dictionary((0, record_1.mapRecord)(this.dictionary, callbackFn));
    }
    get flatten() {
        return this.dictionary;
    }
}
exports.Dictionary = Dictionary;
class DictionaryTypeBuilder {
    builder;
    constructor(builder) {
        this.builder = builder;
    }
    build(value, logger) {
        return new Dictionary((0, record_1.mapRecord)(value, value => this.builder.build(value, logger)));
    }
}
exports.DictionaryTypeBuilder = DictionaryTypeBuilder;
