// firebase-function/lib/src/firestore/FirestoreDocument.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FirestoreDocument = void 0;
const Flattable_1 = require("./../utils/Flattable");
const FirestoreCollection_1 = require("./FirestoreCollection");
const FirestoreSnapshot_1 = require("./FirestoreSnapshot");
class FirestoreDocument {
    firestore;
    path;
    constructor(firestore, path) {
        this.firestore = firestore;
        this.path = path;
    }
    collection(key) {
        return new FirestoreCollection_1.FirestoreCollection(this.firestore, this.path.appending(key));
    }
    async snapshot() {
        const snapshot = await this.firestore.doc(this.path.fullPath).get();
        return new FirestoreSnapshot_1.FirestoreSnapshot(snapshot);
    }
    async set(values) {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
        await this.firestore.doc(this.path.fullPath).set(Flattable_1.Flattable.flatten(values));
    }
    async remove() {
        await this.firestore.doc(this.path.fullPath).delete();
    }
}
exports.FirestoreDocument = FirestoreDocument;
