// firebase-function/lib/src/firestore/FirestorePath.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FirestorePath = void 0;
class FirestorePath {
    components;
    constructor(...components) {
        this.components = components;
    }
    append(...components) {
        for (const component of components) {
            if (component.includes('/'))
                throw new Error(`Invalid component: ${component}`);
            this.components.push(component);
        }
    }
    appending(...components) {
        const path = new FirestorePath(...this.components);
        path.append(...components);
        return path;
    }
    get isEmpty() {
        return this.components.length === 0;
    }
    get parentPath() {
        if (this.components.length === 0)
            throw new Error('Root path has no parent');
        return new FirestorePath(...this.components.slice(0, -1));
    }
    get lastComponent() {
        return this.components[this.components.length - 1];
    }
    get fullPath() {
        return this.components.join('/');
    }
}
exports.FirestorePath = FirestorePath;
